### ISSUE TITLE
Evaluate whether use case needs a new ARIA feature — a way to define page-wide keyboard shortcuts

### ISSUE THREAD

**Issue by aleventhal** [w3c/aria#2351]
JAWS currently supports a non-standard attribute called data-at-shortcutkeys, used by Twitter/X and Facebook. I'm in the process of helping JAWS remove support for anything nonstandard, but they need to be able to support the same features if possible.
A lot of web apps have page-wide shortcut keys (Gmail is an example), so this does seem to be a use case.

How should web pages code this up?
- Should ARIA consider a new feature here? Could we consider reusing aria-keyshortcuts somehow? E.g. allow it with a JSON list of keys on non-focusable container elements? Or just the body?

The format:
- Currently this attribute provides JSON with key/value pairs — "[key name]": "localized description".
- JAWS splits it into key/value pairs and we check to see which key we should allow through to the web page.
- Example from facebook.com (in the actual version quotation marks are escaped with their entity, &quote;):
```
<div data-at-shortcutkeys="{"?":"Show or hide keyboard help","/":"Search Facebook","c":"Comment on current post","enter":"See more of current post","f":"Toggle fullscreen","j":"Go to next post in a feed or next photo in photo viewer","k":"Go to previous post in a feed or previous photo in photo viewer","l":"Like or react to current post in a feed or photo in photo viewer","o":"Open attachment from current post","p":"Create new post","q":"Search Messenger contacts","s":"Share current post"}"></div>
```

**Comment by aleventhal** [comment]
+ @mcking65

**Comment by michaelDCurran** [comment]
NVDA does not currently support the existing non-standard version, however if this were standardized, we may at very least report the fact a web page has custom keys, and perhaps list them on demand.

**Comment by aleventhal** [comment]
I propose that we repurpose aria-keyshortcuts when it's on the <body> of a web page. Anywhere else it acts as it does now.

**Comment by stes-acc** [comment]
> I propose that we repurpose aria-keyshortcuts when it's on the of a web page. Anywhere else it acts as it does now.

On the what? The body?

**Comment by aleventhal** [comment]
> > I propose that we repurpose aria-keyshortcuts when it's on the of a web page. Anywhere else it acts as it does now.
> 
> On the what? The body?

Yes, I forgot the backticks.

**Comment by stes-acc** [comment]
Hm this means that aria-keyshortcuts gonna be allowed on containers (by principle, non-focusable, non-interactive stuff) too.

https://www.w3.org/TR/html-aam-1.0/#el-body says role "generic" which maps for instance to ["group" in UIA](https://www.w3.org/TR/core-aam-1.2/#role-map-generic).

This would be OK according to the spec https://www.w3.org/TR/wai-aria-1.3/#aria-keyshortcuts but will require adjustment for best practices ([for example, ARIA APG](https://www.w3.org/WAI/ARIA/apg/practices/keyboard-interface/#keyassignmentconventionsforcommonfunctions)).

More important, if you propose this, you also need to describe how to satisfy the following ARIA spec condition being part of the aria-keyshortcuts definition:

**_**Authors SHOULD provide a way to expose keyboard shortcuts so that all users can discover them, such as through the use of a tooltip.**_** 

A tooltip on the body is a bit pointless (people are not used to invoke that), so at least a different mechanism/pattern to inform all users have to be proposed, too. Of course it could be injected in the context menu of the body, but there needs to be group consensus on this.

**Comment by scottaohara** [comment]
Is there really a reason we need group consensus over something that is an author should, where what an author should do is going to be heavily dependent on the ui / logical location to provide such hints?

the “such as … a tooltip” was only an example, after all.

**Comment by stes-acc** [comment]
"Is there really a reason we need group consensus "

In case of contradictions, it is. There should be more elaborated examples if the only example given does not fit for a variety of reasons.

**Comment by aleventhal** [comment]
I heard skepticism from a number of screen reader users about the way this is implemented by JAWS, where it lets the keys through (as opposed to a mere announcement which can be done via other means). It would be good to find out whether JAWS still considers this feature important. After all, @mcking65 expressed some misgivings about the way this is done for Facebook. Also, NVDA said they do not plan to implement that way. Maybe the best thing is to axe the feature and possibly replace with JAWS scripts that are aware of Facebook and Twitter, rather than have a piece of standard markup that has high potential for abuse.

**Comment by aleventhal** [comment]
@stes-acc I think we should determine whether we even want the feature first, and only if the answer is yes, determine what mechanism (including whether reusing existing markup makes any sense).

**Comment by MarioBatusic** [comment]
I find the idea of listing global keyboard shortcuts / keys for all users very important and useful. Whether we reuse aria-shortcuts for that or define something new. Concerning the way of displaying this keyboard info I thinkk the best would be to have a browser Menu item / Shortcut for this task. Imlementing an additional SR function for that can either exclude non-scren reader users from this information or double the voice output for SR users if using "title" attribute.

**Comment by cookiecrook** [comment]
> I find the idea of listing global keyboard shortcuts / keys for all users very important and useful. 

Could this be a general web platform feature? OpenUI? WICG? WHATWG?

**Comment by spectranaut** [comment]
Minutes from last weeks meeting: https://www.w3.org/2024/10/17-aria-minutes#0e73

**Comment by MarioBatusic** [comment]
Some important things for our co-work with openUI:
* Browser should list all global keys / keyboard shortcuts, not only single-letter ones. For example, we have a number of global shortcut keys combined with alt or shift key in our cloud solutions.
* Browsers could offer this list in a menu or toolbar. The commands can be listed and executable in the same time.
* Windows screen readers should not let the single-letter-key commands through in their reading mode. After consulting the browser global shortcut list, the user can switch in the focus mode to be able to use the shotcuts.

**Comment by aleventhal** [comment]
@keithamus do I recall correctly that you were planning to bring this up in Open UI or HTML? Or was that someone else? Wanting to get some kind of update.

**Comment by keithamus** [comment]
Yes I'm working on a proposal doc for OpenUI which I'll present to them in the coming weeks. It's a very big topic but I hope to have something this month or next perhaps.
